orbit/pkg/update/notifications.go

```diff
@@ -397,3 +398,86 @@ func (h *runScriptsConfigFetcher) GetConfig() (*fleet.OrbitConfig, error) {
 	}
 	return cfg, err
 }
+
+type BitlockDiskEncryptionPayload interface {
+	SetOrUpdateDiskEncryptionPayload(diskEncryptionStatus fleet.OrbitHostDiskEncryptionKeyPayload) error
+}
+
+type windowsMDMBitlockerConfigFetcher struct {
+	// Fetcher is the OrbitConfigFetcher that will be wrapped. It is responsible
+	// for actually returning the orbit configuration or an error.
+	Fetcher OrbitConfigFetcher
+	// Frequency is the minimum amount of time that must pass between two
+	// executions of the windows MDM enrollment attempt.
+	Frequency time.Duration
+	// HostUUID is the current host's UUID.
+	HostUUID string
+
+	// Bitlocker Operation Results
+	EncryptionResult BitlockDiskEncryptionPayload
+
+	lastEnrollRun time.Time
+}
+
+func ApplyWindowsMDMBitlockerFetcherMiddleware(
+	fetcher OrbitConfigFetcher,
+	frequency time.Duration,
+	hostUUID string,
+	encryptionResult BitlockDiskEncryptionPayload,
+) OrbitConfigFetcher {
+	return &windowsMDMBitlockerConfigFetcher{
+		Fetcher:          fetcher,
+		Frequency:        frequency,
+		HostUUID:         hostUUID,
+		EncryptionResult: encryptionResult,
+	}
+}
+
+// GetConfig calls the wrapped Fetcher's GetConfig method, and if the fleet
+// server set the "needs windows enrollment" flag to true, executes the command
+// to enroll into Windows MDM (or not, if the device is a Windows Server).
+func (w *windowsMDMBitlockerConfigFetcher) GetConfig() (*fleet.OrbitConfig, error) {
+	cfg, err := w.Fetcher.GetConfig()
+
+	if err == nil {
+		if cfg.Notifications.EnforceBitLocker {
+			w.attemptBitlockerEncryption(cfg.Notifications)
+		}
+	}
+	return cfg, err
+}
+
+func (w *windowsMDMBitlockerConfigFetcher) attemptBitlockerEncryption(notifs fleet.OrbitConfigNotifications) {
+	// do not trigger Bitlocker encryption if running on a Windwos server
+	isWindowsServer, err := IsRunningOnWindowsServer()
+	if err != nil || isWindowsServer {
+		log.Debug().Msg("skipped calling RegisterDeviceWithManagement to enroll Windows device, device is a server")
+		return
+	}
+
+	if time.Since(w.lastEnrollRun) <= w.Frequency {
```

Comment:
I think you'll need the lock as we do in the other middlewares, the problem is that we have two config fetchers running simultaneously.

Reply:
yes, I was kind of naive about not including it, mostly because there was only one config value being gathered. I missed the fact that two config fetchers run at the same time

Reply:
Yeah you'll need a mutex, you can look at e.g. the Windows MDM enrollment config fetcher, it should be similar.

Reply:
done - change is [here](https://github.com/fleetdm/fleet/pull/14022/files#diff-b5459f59793c455e03b32f0ae9f6a52334f0fc85dcb14ba4402318af7348e77fR444)